common/txmgr/abandoned_tracker.go

```diff
@@ -0,0 +1,194 @@
+	return AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]{
+		txStore: txStore,
+		ks:      keystore,
+		chainID: chainID,
+		client:  client,
+		lggr:    lggr.Named("AbandonedTracker"),
+		txes:    make([]AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE], 0),
+	}
+}
+
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) containsTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+	for _, atx := range tracker.txes {
+		if atx.tx.ID == tx.ID {
+			return true
+		}
+	}
+	return false
+}
+
+// insertAbandonedTx inserts a transaction into the tracker
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) insertAbandonedTx(
+	tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	if tracker.containsTx(tx) {
+		return
+	}
+
+	tracker.lggr.Debugw(fmt.Sprintf("inserting tx %v", tx.ID))
+	tracker.txes = append(tracker.txes, AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]{
+		tx:        tx,
+		fatalTime: time.Now().Add(TTL),
+	})
+}
+
+// markFatal sets a transaction's state to fatal_error
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) markFatal(
+	ctx context.Context,
+	atx AbandonedTx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) {
+	tracker.lggr.Infow(fmt.Sprintf("tx %v marked as fatal for exceeding ttl", atx.tx.ID))
+
+	atx.tx.Error.SetValid(AbandonedErrorMsg)
+
+	err := (*tracker.txStore).UpdateTxFatalError(ctx, atx.tx)
+	if err != nil {
+		tracker.lggr.Errorw(fmt.Sprintf("failed to mark tx %v as fatal", atx.tx.ID))
+		// TODO: Handle error
+	}
+}
+
+// getAbandonedAddresses retrieves fromAddress’s in evm.txes that are not present in the Confirmer's enabledAddresses list
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) getAbandonedAddresses(enabledAddrs []ADDR) ([]ADDR, error) {
+	fromAddresses, err := (*tracker.ks).EnabledAddressesForChain(tracker.chainID)
+	if err != nil {
+		return nil, err
+	}
+
+	var abandoned []ADDR
+	for _, addr := range fromAddresses {
+		if !slices.Contains(enabledAddrs, addr) {
+			abandoned = append(abandoned, addr)
+		}
+	}
+
+	return abandoned, nil
+}
+
+// HandleAbandonedTransactions is called by the Confirmer to track and handle all abandoned transactions
+func (tracker *AbandonedTracker[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) HandleAbandonedTransactions(ctx context.Context, enabledAddrs []ADDR) {
+	// Find abandoned addresses
+	abandonedAddrs, err := tracker.getAbandonedAddresses(enabledAddrs)
+	if err != nil {
+		// TODO handle error
+	}
+
+	// Get abandoned txes from addresses and insert into the tracker
+	for _, addr := range abandonedAddrs {
+		seq, err := (*tracker.client).SequenceAt(ctx, addr, nil)
+		if err != nil {
+			// TODO handle error
+		}
+
+		tx, err := (*tracker.txStore).FindTxWithSequence(ctx, addr, seq)
+		if err != nil {
+			// TODO handle error
+		}
```

Comment:
I think getting the tx's like this would have some problem. A tx in our DB may not have a Sequence/Nonce associated with it until it's ready to be broadcasted. Using the sequence to retrieve tx wouldn't capture all of the tx's we want to track here. If we pull the addresses like in my previous comment, we might not need this loop anyways.

Reply:
We may not need `tracket.txes` either if we decide to pull the list of tx from the DB every time. But idk if that's the best approach either depending how often `processHead` is called